File: java/src/main/java/it/giannibombelli/wsc2026/booking/domain/policies/BookingPaymentRequestPolicy.java
package it.giannibombelli.wsc2026.booking.domain.policies;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.events.BookingPlaced;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.model.Policy;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.application.commands.RequestPayment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;


public class BookingPaymentRequestPolicy implements Policy<BookingPlaced, RequestPayment> {

    @Override
    public RequestPayment evaluate(BookingPlaced event) {
        Require.requireArgument(event, "Booking placed event");
        return new RequestPayment(
            EntityId.generate(PaymentId::new),
            new ClientReference(event.aggregateId().value().toString()),
            event.amount(),
            Timestamp.now()
        );
    }
}

File: typescript/src/booking/domain/policies/bookingPaymentRequestPolicy.ts
// Policy che traduce un BookingPlaced in una richiesta di pagamento per il Payment BC.


import type { Policy } from '@/common/domain/model/policy.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { RequestPayment, requestPayment } from '@/payment/application/commands/requestPayment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { BookingPlaced } from '../events/bookingPlaced.js';

export class BookingPaymentRequestPolicy implements Policy<BookingPlaced, RequestPayment> {
  evaluate(event: BookingPlaced): RequestPayment {
    requireArgument(event, 'Booking placed event');
    return requestPayment(
      generateId((value) => new PaymentId(value)),
      new ClientReference(event.aggregateId.value),
      event.amount,
      Timestamp.now()
    );
  }
}
